cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -291,22 +309,36 @@ def method_with_logging(
         func = cls_or_func
 
         @functools.wraps(func)
-        def func_with_logging(
-            circuit: 'cirq.AbstractCircuit', context: TransformerContext
-        ) -> 'cirq.AbstractCircuit':
-            return _transform_and_log(func, func.__name__, circuit, context)
+        def func_with_logging(circuit, **kwargs) -> 'cirq.AbstractCircuit':
+            return _transform_and_log(
+                func,
+                func.__name__,
+                circuit,
+                _get_context(func, f'{func.__name__}()', **kwargs),
+                **kwargs,
+            )
 
         return func_with_logging
 
 
+def _get_context(func, name, **kwargs):
+    if 'context' in kwargs:
+        return kwargs['context']
+    sig = inspect.signature(func)
+    default_context = sig.parameters["context"].default
```

Comment:
The protocol has a default value of `None` for `context`. Is it useful to allow people to define transformers that _don't_ have default values for context? If not, we could make this required and then check the default value once when applying the transformer decorator:
```python
func = cls_or_func
default_context = _get_default_context(func)

@functools.wraps(func)
def func_with_logging(circuit, **kwargs):
    return _transform_and_log(func, func.__name__, circuit, kwargs.get('context', default_context), **kwargs)
```

Reply:
> Is it useful to allow people to define transformers that don't have default values for context?

No, and given the modified `Transformer` protocol, a decorated class/method without a default value for context will fail mypy type check

> If not, we could make this required and then check the default value once when applying the transformer decorator:

Good suggestion, done!